src/platform/OpenThread/GenericThreadBorderRouterDelegate.cpp

```diff
@@ -0,0 +1,193 @@
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
+    otBorderAgentId borderAgentId;
+    if (borderAgentIdSpan.size() < sizeof(borderAgentId.mId))
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    ScopedThreadLock threadLock;
+    otError err = otBorderAgentGetId(otInst, &borderAgentId);
+    if (err == OT_ERROR_NONE)
+    {
+        memcpy(borderAgentIdSpan.data(), borderAgentId.mId, sizeof(borderAgentId.mId));
+        borderAgentIdSpan.reduce_size(sizeof(borderAgentId.mId));
+        return CHIP_NO_ERROR;
+    }
+    return CHIP_ERROR_INTERNAL;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetThreadVersion(uint16_t & threadVersion)
+{
+    threadVersion = otThreadGetVersion();
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetInterfaceEnabled(bool & interfaceEnabled)
+{
+    otInstance * otInst = DeviceLayer::ThreadStackMgrImpl().OTInstance();
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
+    ScopedThreadLock threadLock;
+    interfaceEnabled = otIp6IsEnabled(otInst);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::GetDataset(Thread::OperationalDataset & dataset, DatasetType type)
+{
+    otInstance * otInst = DeviceLayer::ThreadStackMgrImpl().OTInstance();
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
+
+    ScopedThreadLock threadLock;
+    otError otErr = OT_ERROR_NONE;
+    otOperationalDatasetTlvs datasetTlvs;
+    if (type == DatasetType::kActive)
+    {
+        otErr = otDatasetGetActiveTlvs(otInst, &datasetTlvs);
+    }
+    else
+    {
+        otErr = otDatasetGetPendingTlvs(otInst, &datasetTlvs);
+    }
+    if (otErr == OT_ERROR_NONE)
+    {
+        return dataset.Init(ByteSpan(datasetTlvs.mTlvs, datasetTlvs.mLength));
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR GenericOpenThreadBorderRouterDelegate::SetActiveDataset(const Thread::OperationalDataset & activeDataset,
+                                                                   uint32_t sequenceNum, ActivateDatasetCallback * callback)
+{
+    // This function will never be invoked when there is an Active Dataset already configured.
+    CHIP_ERROR err = SaveThreadBorderRouterCommissioned(false);
+    if (err == CHIP_NO_ERROR)
+    {
+        err = DeviceLayer::ThreadStackMgrImpl().AttachToThreadNetwork(activeDataset, nullptr);
+    }
+    if (err == CHIP_NO_ERROR)
+    {
+        // We can change the stored sequence number because OnActivateDatasetComplete executed later must be called for this
+        // SetActiveDataset request
+        mSequenceNum = sequenceNum;
+        mCallback    = callback;
+    }
+    return err;
+}
+
+void GenericOpenThreadBorderRouterDelegate::OnPlatformEventHandler(const DeviceLayer::ChipDeviceEvent * event, intptr_t arg)
+{
+    GenericOpenThreadBorderRouterDelegate * delegate = reinterpret_cast<GenericOpenThreadBorderRouterDelegate *>(arg);
+    if (delegate && delegate->mCallback)
```

Comment:
What happens if either is null? Is it possible a callback will never be called?

Reply:
If the thread br is attached to a Thread network without the help of thread br management cluster(it can use cli or other way to do this) then the mCallback will be null.
It is possible that a callback is never called by this function. It might be: 1. client sends SetActiveDatasetRequest and the  TBR starts to attach to the Thread network. 2. Failsafe timer expires before the TBR attaches to the Thread network, the TBR will call the callback in FailSafe expired event handler with timeout error but not in this function.